File: app/src/main/java/com/pr0gramm/app/ui/MenuSheetView.java
package com.pr0gramm.app.ui;

import android.annotation.SuppressLint;
import android.annotation.TargetApi;
import android.content.Context;
import android.content.res.ColorStateList;
import android.graphics.Color;
import android.graphics.Outline;
import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.support.annotation.ColorInt;
import android.support.annotation.MenuRes;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.support.v4.view.ViewCompat;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.SubMenu;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewOutlineProvider;
import android.widget.AbsListView;
import android.widget.BaseAdapter;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.PopupMenu;
import android.widget.TextView;

import com.pr0gramm.app.R;
import com.pr0gramm.app.util.AndroidUtility;

import java.util.ArrayList;

/**
 * A SheetView that can represent a menu resource as a list or grid.
 * <p>
 * A list can support submenus, and will include a divider and header for them where appropriate.
 * Grids currently don't support submenus, and don't in the Material Design spec either.
 * </p>
 */
@SuppressLint("ViewConstructor")
public class MenuSheetView extends FrameLayout {
    @ColorInt
    private final int defaultColor = ColorStateList.valueOf(Color.BLACK)
            .withAlpha(127)
            .getDefaultColor();

    /**
     * A listener for menu item clicks in the sheet
     */
    public interface OnMenuItemClickListener {
        boolean onMenuItemClick(MenuItem item);
    }

    private Menu menu;
    private ArrayList<SheetMenuItem> items = new ArrayList<>();
    private Adapter adapter;
    private AbsListView absListView;
    private final TextView titleView;
    protected final int originalListPaddingTop;

    /**
     * @param context  Context to construct the view with
     * @param titleRes String resource ID for the title
     * @param listener Listener for menu item clicks in the sheet
     */
    public MenuSheetView(final Context context, @StringRes final int titleRes, final OnMenuItemClickListener listener) {
        this(context, context.getString(titleRes), listener);
    }

    /**
     * @param context  Context to construct the view with
     * @param title    Title for the sheet. Can be null
     * @param listener Listener for menu item clicks in the sheet
     */
    public MenuSheetView(final Context context, @Nullable final CharSequence title, final OnMenuItemClickListener listener) {
        super(context);

        // Set up the menu
        this.menu = new PopupMenu(context, null).getMenu();  // Dirty hack to get a menu instance since MenuBuilder isn't public ಠ_ಠ

        // Inflate the appropriate view and set up the absListView
        inflate(context, R.layout.list_sheet_view, this);
        absListView = (AbsListView) findViewById(R.id.list);
        if (listener != null) {
            absListView.setOnItemClickListener((parent, view, position, id) ->
                    listener.onMenuItemClick(adapter.getItem(position).getMenuItem()));
        }

        // Set up the title
        titleView = (TextView) findViewById(R.id.title);
        originalListPaddingTop = absListView.getPaddingTop();
        setTitle(title);

        ViewCompat.setElevation(this, AndroidUtility.dp(getContext(), 16));
    }

    /**
     * Inflates a menu resource into the menu backing this sheet.
     *
     * @param menuRes Menu resource ID
     */
    public void inflateMenu(@MenuRes int menuRes) {
        if (menuRes != -1) {
            MenuInflater inflater = new MenuInflater(getContext());
            inflater.inflate(menuRes, menu);
        }

        prepareMenuItems();
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
        this.adapter = new Adapter();
        absListView.setAdapter(this.adapter);
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        // Necessary for showing elevation on 5.0+
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            setOutlineProvider(new ShadowOutline(w, h));
        }
    }

    /**
     * A helper class for providing a shadow on sheets
     */
    @TargetApi(21)
    static class ShadowOutline extends ViewOutlineProvider {
        private final int width;
        private final int height;

        ShadowOutline(int width, int height) {
            this.width = width;
            this.height = height;
        }

        @Override
        public void getOutline(View view, Outline outline) {
            outline.setRect(0, 0, width, height);
        }
    }

    /**
     * Flattens the visible menu items of {@link #menu} into {@link #items},
     * while inserting separators between items when necessary.
     * <p>
     * Adapted from the Design support library's NavigationMenuPresenter implementation
     */
    private void prepareMenuItems() {
        items.clear();
        int currentGroupId = 0;

        // Iterate over the menu items
        for (int i = 0; i < menu.size(); i++) {
            MenuItem item = menu.getItem(i);
            if (item.isVisible()) {
                if (item.hasSubMenu()) {
                    // Flatten the submenu
                    SubMenu subMenu = item.getSubMenu();
                    if (subMenu.hasVisibleItems()) {
                        items.add(SheetMenuItem.SEPARATOR);

                        // Add a header item if it has text
                        if (!TextUtils.isEmpty(item.getTitle())) {
                            items.add(SheetMenuItem.of(item));
                        }

                        // Add the sub-items
                        for (int subI = 0, size = subMenu.size(); subI < size; subI++) {
                            MenuItem subMenuItem = subMenu.getItem(subI);
                            if (subMenuItem.isVisible()) {
                                items.add(SheetMenuItem.of(subMenuItem));
                            }
                        }

                        // Add one more separator to the end to close it off if we have more items
                        if (i != menu.size() - 1) {
                            items.add(SheetMenuItem.SEPARATOR);
                        }
                    }
                } else {
                    int groupId = item.getGroupId();
                    if (groupId != currentGroupId) {
                        items.add(SheetMenuItem.SEPARATOR);
                    }
                    items.add(SheetMenuItem.of(item));
                    currentGroupId = groupId;
                }
            }
        }
    }

    /**
     * @return The current {@link Menu} instance backing this sheet. Note that this is mutable, and
     * you should call {@link #updateMenu()} after any changes.
     */
    public Menu getMenu() {
        return this.menu;
    }

    /**
     * Invalidates the current internal representation of the menu and rebuilds it. Should be used
     * if the developer dynamically adds items to the Menu returned by {@link #getMenu()}
     */
    public void updateMenu() {
        // Invalidate menu and rebuild it, useful if the user has dynamically added menu items.
        prepareMenuItems();
    }

    /**
     * Sets the title text of the sheet
     *
     * @param resId String resource ID for the text
     */
    public void setTitle(@StringRes int resId) {
        setTitle(getResources().getText(resId));
    }

    /**
     * Sets the title text of the sheet
     *
     * @param title Title text to use
     */
    public void setTitle(CharSequence title) {
        if (!TextUtils.isEmpty(title)) {
            titleView.setText(title);
        } else {
            titleView.setVisibility(GONE);

            // Add some padding to the top to account for the missing title
            absListView.setPadding(absListView.getPaddingLeft(),
                    originalListPaddingTop + AndroidUtility.dp(getContext(), 8),
                    absListView.getPaddingRight(), absListView.getPaddingBottom());
        }
    }

    /**
     * @return The current title text of the sheet
     */
    public CharSequence getTitle() {
        return titleView.getText();
    }

    private class Adapter extends BaseAdapter {

        private static final int VIEW_TYPE_NORMAL = 0;
        private static final int VIEW_TYPE_SUBHEADER = 1;
        private static final int VIEW_TYPE_SEPARATOR = 2;

        private final LayoutInflater inflater;

        public Adapter() {
            this.inflater = LayoutInflater.from(getContext());
        }

        @Override
        public int getCount() {
            return items.size();
        }

        @Override
        public SheetMenuItem getItem(int position) {
            return items.get(position);
        }

        @Override
        public long getItemId(int position) {
            return position;
        }

        @Override
        public int getViewTypeCount() {
            return 3;
        }

        @Override
        public int getItemViewType(int position) {
            SheetMenuItem item = getItem(position);
            if (item.isSeparator()) {
                return VIEW_TYPE_SEPARATOR;
            } else if (item.getMenuItem().hasSubMenu()) {
                return VIEW_TYPE_SUBHEADER;
            } else {
                return VIEW_TYPE_NORMAL;
            }
        }

        @Override
        public View getView(int position, View convertView, ViewGroup parent) {

            SheetMenuItem item = getItem(position);
            int viewType = getItemViewType(position);

            switch (viewType) {
                case VIEW_TYPE_NORMAL:
                    NormalViewHolder holder;
                    if (convertView == null) {
                        convertView = inflater.inflate(R.layout.sheet_list_item, parent, false);
                        holder = new NormalViewHolder(convertView);
                        convertView.setTag(holder);
                    } else {
                        holder = (NormalViewHolder) convertView.getTag();
                    }
                    holder.bindView(item);
                    break;
                case VIEW_TYPE_SUBHEADER:
                    if (convertView == null) {
                        convertView = inflater.inflate(R.layout.sheet_list_item_subheader, parent, false);
                    }
                    TextView subHeader = (TextView) convertView;
                    subHeader.setText(item.getMenuItem().getTitle());
                    break;
                case VIEW_TYPE_SEPARATOR:
                    if (convertView == null) {
                        convertView = inflater.inflate(R.layout.sheet_list_item_separator, parent, false);
                    }
                    break;
            }

            return convertView;
        }

        @Override
        public boolean areAllItemsEnabled() {
            return false;
        }

        @Override
        public boolean isEnabled(int position) {
            return getItem(position).isEnabled();
        }

        class NormalViewHolder {
            final ImageView icon;
            final TextView label;

            NormalViewHolder(View root) {
                icon = (ImageView) root.findViewById(R.id.icon);
                label = (TextView) root.findViewById(R.id.label);
            }

            public void bindView(SheetMenuItem item) {
                Drawable icon = item.getMenuItem().getIcon();
                icon.mutate().setColorFilter(defaultColor, PorterDuff.Mode.SRC_IN);

                this.icon.setImageDrawable(icon);
                label.setText(item.getMenuItem().getTitle());
            }
        }
    }

    private static class SheetMenuItem {

        private static final SheetMenuItem SEPARATOR = new SheetMenuItem(null);

        private final MenuItem menuItem;

        private SheetMenuItem(MenuItem item) {
            menuItem = item;
        }

        public static SheetMenuItem of(MenuItem item) {
            return new SheetMenuItem(item);
        }

        public boolean isSeparator() {
            return this == SEPARATOR;
        }

        public MenuItem getMenuItem() {
            return menuItem;
        }

        public boolean isEnabled() {
            // Separators and subheaders never respond to click
            return menuItem != null && !menuItem.hasSubMenu() && menuItem.isEnabled();
        }

    }
}

File: app/src/main/java/com/pr0gramm/app/ui/MenuSheetView.kt
package com.pr0gramm.app.ui

import android.annotation.SuppressLint
import android.annotation.TargetApi
import android.content.Context
import android.graphics.Outline
import android.graphics.PorterDuff
import android.text.TextUtils
import android.view.*
import android.widget.*
import androidx.annotation.MenuRes
import androidx.annotation.StringRes
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import com.pr0gramm.app.R
import com.pr0gramm.app.util.AndroidUtility
import com.pr0gramm.app.util.dip2px
import com.pr0gramm.app.util.find
import java.util.*

typealias OnMenuItemClickListener = (MenuItem) -> Unit

/**
 * A SheetView that can represent a menu resource as a list.
 *
 *
 * A list can support submenus, and will include a divider and header for them where appropriate.
 *
 */
@SuppressLint("ViewConstructor")
class MenuSheetView(context: Context, @StringRes titleRes: Int, listener: OnMenuItemClickListener) : FrameLayout(context) {
    private val menu: Menu = androidx.appcompat.widget.PopupMenu(context, this).menu

    private val items = ArrayList<SheetMenuItem>()
    private val absListView: AbsListView
    private val titleView: TextView
    private val originalListPaddingTop: Int

    init {
        // Set up the menu
        // Dirty hack to get a menu instance since MenuBuilder isn't public ಠ_ಠ

        // Inflate the appropriate view and set up the absListView
        View.inflate(context, R.layout.list_sheet_view, this)
        absListView = findViewById<View>(R.id.list) as AbsListView
        absListView.onItemClickListener = AdapterView.OnItemClickListener { _, _, position, id ->
            val item = absListView.adapter.getItem(position) as? SheetMenuItem
            if (item?.menuItem != null) {
                listener(item.menuItem)
            }
        }

        // Set up the title
        titleView = findViewById<View>(R.id.title) as TextView
        originalListPaddingTop = absListView.paddingTop
        setTitle(context.getString(titleRes))

        ViewCompat.setElevation(this, getContext().dip2px(16).toFloat())
    }

    /**
     * Inflates a menu resource into the menu backing this sheet.
     *
     * @param menuRes Menu resource ID
     */
    fun inflateMenu(@MenuRes menuRes: Int) {
        if (menuRes != -1) {
            val inflater = (context as AppCompatActivity).menuInflater
            inflater.inflate(menuRes, menu)
        }

        prepareMenuItems()
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        absListView.adapter = Adapter()
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        outlineProvider = ShadowOutline(w, h)
    }

    /**
     * A helper class for providing a shadow on sheets
     */
    @TargetApi(21)
    internal class ShadowOutline(private val width: Int, private val height: Int) : ViewOutlineProvider() {
        override fun getOutline(view: View, outline: Outline) {
            outline.setRect(0, 0, width, height)
        }
    }

    /**
     * Flattens the visible menu items of [.menu] into [.items],
     * while inserting separators between items when necessary.
     *
     *
     * Adapted from the Design support library's NavigationMenuPresenter implementation
     */
    private fun prepareMenuItems() {
        items.clear()
        var currentGroupId = 0

        // Iterate over the menu items
        for (i in 0 until menu.size()) {
            val item = menu.getItem(i)
            if (item.isVisible) {
                if (item.hasSubMenu()) {
                    // Flatten the submenu
                    val subMenu = item.subMenu
                    if (subMenu.hasVisibleItems()) {
                        items.add(SheetMenuItem.SEPARATOR)

                        // Add a header item if it has text
                        if (!TextUtils.isEmpty(item.title)) {
                            items.add(SheetMenuItem.of(item))
                        }

                        // Add the sub-items
                        var subI = 0
                        val size = subMenu.size()
                        while (subI < size) {
                            val subMenuItem = subMenu.getItem(subI)
                            if (subMenuItem.isVisible) {
                                items.add(SheetMenuItem.of(subMenuItem))
                            }
                            subI++
                        }

                        // Add one more separator to the end to close it off if we have more items
                        if (i != menu.size() - 1) {
                            items.add(SheetMenuItem.SEPARATOR)
                        }
                    }
                } else {
                    val groupId = item.groupId
                    if (groupId != currentGroupId) {
                        items.add(SheetMenuItem.SEPARATOR)
                    }
                    items.add(SheetMenuItem.of(item))
                    currentGroupId = groupId
                }
            }
        }
    }

    /**
     * Sets the title text of the sheet
     *
     * @param title Title text to use
     */
    private fun setTitle(title: CharSequence?) {
        if (!TextUtils.isEmpty(title)) {
            titleView.text = title
        } else {
            titleView.visibility = View.GONE

            // Add some padding to the top to account for the missing title
            absListView.setPadding(absListView.paddingLeft,
                    originalListPaddingTop + context.dip2px(8),
                    absListView.paddingRight, absListView.paddingBottom)
        }
    }

    private inner class Adapter : BaseAdapter() {
        private val inflater: LayoutInflater = LayoutInflater.from(context)

        private val VIEW_TYPE_NORMAL = 0
        private val VIEW_TYPE_SUBHEADER = 1
        private val VIEW_TYPE_SEPARATOR = 2

        override fun getCount(): Int {
            return items.size
        }

        override fun getItem(position: Int): SheetMenuItem {
            return items[position]
        }

        override fun getItemId(position: Int): Long {
            return position.toLong()
        }

        override fun getViewTypeCount(): Int {
            return 3
        }

        override fun getItemViewType(position: Int): Int {
            val item = getItem(position)
            return if (item.isSeparator) {
                VIEW_TYPE_SEPARATOR
            } else if (item.menuItem!!.hasSubMenu()) {
                VIEW_TYPE_SUBHEADER
            } else {
                VIEW_TYPE_NORMAL
            }
        }

        override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
            var view = convertView

            val item = getItem(position)
            val viewType = getItemViewType(position)

            when (viewType) {
                VIEW_TYPE_NORMAL -> {
                    val holder: NormalViewHolder
                    if (view == null) {
                        view = inflater.inflate(R.layout.sheet_list_item, parent, false)
                        holder = NormalViewHolder(view!!)
                        view.tag = holder
                    } else {
                        holder = view.tag as NormalViewHolder
                    }
                    holder.bindView(item)
                }
                VIEW_TYPE_SUBHEADER -> {
                    if (view == null) {
                        view = inflater.inflate(R.layout.sheet_list_item_subheader, parent, false)
                    }
                    val subHeader = view as TextView?
                    subHeader!!.text = item.menuItem!!.title
                }
                VIEW_TYPE_SEPARATOR -> if (view == null) {
                    view = inflater.inflate(R.layout.sheet_list_item_separator, parent, false)
                }
            }

            return view!!
        }

        override fun areAllItemsEnabled(): Boolean {
            return false
        }

        override fun isEnabled(position: Int): Boolean {
            return getItem(position).isEnabled
        }

        internal inner class NormalViewHolder(root: View) {
            val icon: ImageView = root.find(R.id.icon) as ImageView
            val label: TextView = root.find(R.id.label) as TextView

            fun bindView(item: SheetMenuItem) {
                val iconColor = AndroidUtility.resolveColorAttribute(icon.context,
                        android.R.attr.textColorSecondary)

                val menuItem = item.menuItem ?: return

                menuItem.icon?.let { icon ->
                    icon.mutate().setColorFilter(iconColor, PorterDuff.Mode.SRC_IN)
                    this.icon.setImageDrawable(icon)
                }

                label.text = menuItem.title
            }
        }
    }

    private class SheetMenuItem private constructor(val menuItem: MenuItem?) {

        val isSeparator: Boolean
            get() = this === SEPARATOR

        // Separators and subheaders never respond to click
        val isEnabled: Boolean
            get() = menuItem != null && !menuItem.hasSubMenu() && menuItem.isEnabled

        companion object {

            internal val SEPARATOR = SheetMenuItem(null)

            fun of(item: MenuItem): SheetMenuItem {
                return SheetMenuItem(item)
            }
        }

    }
}
